### PR TITLE
Run blackbox containers preferentially on invoker subset

### PR DESCRIPTION
**Pull request by perryibm**

### CODE REVIEW

**Comment by perryibm** [comment]
PG2-590 passed.

**Comment by markusthoemmes** [comment]
Any tests possible to proof it works as expected?

**Comment by perryibm** [comment]
I examined the above PG which is fine but given the small size of the deployment does not confirm a lot.  My plan was to examine the behavior on the larger deployment.

File: common/scala/src/main/scala/whisk/common/Config.scala [hunk 1]
```diff
@@ -68,6 +68,16 @@ class Config(requiredProperties: Map[String, String], optionalProperties: Set[St
     }
 
     /**
+     * Returns the value of a given key parsed as a double.
+     * If parsing fails, return the default alue.
```
Comment (rabbah) [inline common/scala/src/main/scala/whisk/common/Config.scala]: alue -> value.

File: core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala [hunk 2]
```diff
@@ -297,18 +298,15 @@ class LoadBalancerService(config: WhiskConfig,
                 0
             }
         }
-        //info(this, s"getInvoker: ${subject} ${path} -> ${hash} ${count}")
         return (hash, count)
     }
 
-    private def numInvokers = invokerHealth.getCurStatus.length
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala]: nice.

File: core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala [hunk 3]
```diff
@@ -89,8 +87,8 @@ class LoadBalancerService(config: WhiskConfig,
 
     private implicit val emitter: PrintStreamEmitter = this
 
-    /** How many invokers are dedicated to blackbox images */
-    private val blackboxFraction : Double = config.controllerBlackboxFraction
+    /** How many invokers are dedicated to blackbox images.  We range bound to something sensical regardless of configuration. */
+    private val blackboxFraction : Double = Math.min(0.0, Math.max(1.0, config.controllerBlackboxFraction))
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala]: isn't this always 0?

File: core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala [hunk 4]
```diff
@@ -241,38 +240,44 @@ class LoadBalancerService(config: WhiskConfig,
...
     private def getActiveCountByInvoker(): Map[Int, Long] = activationByInvoker.toMap mapValues { _.size.toLong }
 
-    // A count of how many activations have been posted to Kafka based on invoker index or user/subject.
+    /** A count of how many activations have been posted to Kafka based on invoker index or user/subject. */
     private val userActivationCounter = new TrieMap[String, Counter]
 
+    /** Return a sorted list of available invokers. */
+    private def getAvailableInvokers(): Array[Int] = invokerHealth.getCurStatus.filter( _.isUp).sortBy(_.index).map(_.index)
+
+    /** Compute the number of blackbox-dedicted invokers by applying a rounded down fraction of all invokers (but at least 1). */
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala]: dedicted -> dedicated

File: core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala [hunk 5]
```diff
@@ -241,38 +240,44 @@ class LoadBalancerService(config: WhiskConfig,
...
+        val allInvokers = getAvailableInvokers
+        val numManaged = Math.max(1, allInvokers.length - numBlackbox(allInvokers.length))
+        allInvokers.take(numManaged)
+    }
 
     /**
      * Determine which invoker this activation should go to.  Due to dynamic conditions, it may return no invoker.
      */
-    private def getInvoker(msg: ActivationMessage): Option[Int] = {
+    private def getInvoker(action: WhiskAction, msg: ActivationMessage): Option[Int] = {
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala]: does it make sense to change the action type to just `Exec`?

**Comment by perryibm** [comment]
Rebased.  Fixed typos.  Re-trying PG2-776.

**Comment by rabbah** [comment]
Merged as https://github.com/openwhisk/openwhisk/compare/d191985b6713...3921fde9b488.

**Pull request closed without merging**